src/passes/analysis/verilog.cpp

```diff
@@ -1594,14 +1600,60 @@ bool Passes::Verilog::runOnInstanceGraphNode(InstanceGraphNode& node) {
   return false;
 }
 
+bool add_prefix(
+  std::unique_ptr<vAST::AbstractModule>& module,
+  std::string prefix) {
+  // Note we cannot add prefix to string modules since their
+  // module name is inside an opaque verilog string
+  if (auto ptr = dynamic_cast<vAST::Module*>(module.get())) {
+    ptr->name = prefix + ptr->name;
+    return true;
+  }
+  return false;
+}
+
+class InstancePrefixInserter : public vAST::Transformer {
+  std::set<std::string> renamed_modules;
+  std::string prefix;
+
+ public:
+  InstancePrefixInserter(
+    std::set<std::string> renamed_modules,
+    std::string prefix)
+      : renamed_modules(renamed_modules),
+        prefix(prefix){};
+
+  using vAST::Transformer::visit;
+  virtual std::unique_ptr<vAST::ModuleInstantiation> visit(
+    std::unique_ptr<vAST::ModuleInstantiation> node) {
+    if (renamed_modules.count(node->module_name)) {
+      node->module_name = prefix + node->module_name;
+    }
+    return node;
+  }
+};
+
 void Passes::Verilog::writeToStream(std::ostream& os) {
```

Comment:
why do this in `writeToStream`? I think we should do this in the main run function.

Reply:
Which run function are you referring to? We can't do this on runOnInstanceGraphNode since we need to do a "global" pass to see which modules are renamed since not all are, so we need to do this before or after the traversal.  I chose to do it here since at this point, the modules are all generated and we are simply traversing them and inserting the prefix before the final step in code generation, which seemed like a reasonable place to do it.

Reply:
ugh yeah I see, really the passes should have a `finalize()` hook, but doesn't look like it's in the infra right now. problem is that `writeToFiles()` also needs this logic. I would suggest encapsulating this as follows, (assuming `b ool prefixAdded = False;` is in the class declaration in the hpp)

```cpp
void Passes::Verilog::addPrefix() {
    if (prefixAdded || module_name_prefix == "") return;
    // Renaming logic here...
    for (auto& module : modules) {
        ...
    }
    prefixAdded = True;
}

 void Passes::Verilog::writeToStream(std::ostream& os) {
    addPrefix();
    // Rest of logic here...
}

 void Passes::Verilog::writeToFiles(...) {
    addPrefix();
    // Rest of logic here...
}
```